source/loader/layers/sanitizer/ur_sanddi.cpp

```diff
@@ -0,0 +1,481 @@
+) {
+    if (nullptr == pDdiTable) {
+        return UR_RESULT_ERROR_INVALID_NULL_POINTER;
+    }
+
+    if (UR_MAJOR_VERSION(ur_sanitizer_layer::context.version) !=
+            UR_MAJOR_VERSION(version) ||
+        UR_MINOR_VERSION(ur_sanitizer_layer::context.version) >
+            UR_MINOR_VERSION(version)) {
+        return UR_RESULT_ERROR_UNSUPPORTED_VERSION;
+    }
+
+    ur_result_t result = UR_RESULT_SUCCESS;
+
+    pDdiTable->pfnCreate = ur_sanitizer_layer::urQueueCreate;
+    pDdiTable->pfnRelease = ur_sanitizer_layer::urQueueRelease;
+
+    return result;
+}
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Exported function for filling application's USM table
+///        with current process' addresses
+///
+/// @returns
+///     - ::UR_RESULT_SUCCESS
+///     - ::UR_RESULT_ERROR_INVALID_NULL_POINTER
+///     - ::UR_RESULT_ERROR_UNSUPPORTED_VERSION
+__urdlllocal ur_result_t UR_APICALL urGetUSMProcAddrTable(
+    ur_api_version_t version, ///< [in] API version requested
+    ur_usm_dditable_t
+        *pDdiTable ///< [in,out] pointer to table of DDI function pointers
+) {
+    if (nullptr == pDdiTable) {
+        return UR_RESULT_ERROR_INVALID_NULL_POINTER;
+    }
+
+    if (UR_MAJOR_VERSION(ur_sanitizer_layer::context.version) !=
+            UR_MAJOR_VERSION(version) ||
+        UR_MINOR_VERSION(ur_sanitizer_layer::context.version) >
+            UR_MINOR_VERSION(version)) {
+        return UR_RESULT_ERROR_UNSUPPORTED_VERSION;
+    }
+
+    ur_result_t result = UR_RESULT_SUCCESS;
+
+    pDdiTable->pfnDeviceAlloc = ur_sanitizer_layer::urUSMDeviceAlloc;
+
+    return result;
+}
+
+ur_result_t context_t::init(ur_dditable_t *dditable,
+                            const std::set<std::string> &enabledLayerNames,
+                            [[maybe_unused]] codeloc_data codelocData) {
+    ur_result_t result = UR_RESULT_SUCCESS;
+
+    if (enabledLayerNames.count("UR_LAYER_ASAN")) {
+        context.enabledType = SanitizerType::AddressSanitizer;
+    } else if (enabledLayerNames.count("UR_LAYER_MSAN")) {
+        context.enabledType = SanitizerType::MemorySanitizer;
+    } else if (enabledLayerNames.count("UR_LAYER_TSAN")) {
+        context.enabledType = SanitizerType::ThreadSanitizer;
+    }
+
+    // Only support AddressSanitizer now
+    if (context.enabledType != SanitizerType::AddressSanitizer) {
+        return result;
+    }
+
+    if (context.enabledType == SanitizerType::AddressSanitizer) {
+        if (!(dditable->VirtualMem.pfnReserve && dditable->VirtualMem.pfnMap &&
+              dditable->VirtualMem.pfnGranularityGetInfo)) {
+            die("Some VirtualMem APIs are needed to enable UR_LAYER_ASAN");
+        }
+
+        if (!dditable->PhysicalMem.pfnCreate) {
+            die("Some PhysicalMem APIs are needed to enable UR_LAYER_ASAN");
+        }
+    }
+
+    // FIXME: Interceptor needs use some of APIs that aren't registered, maybe just copy needed APIs?
```

Comment:
based on the discussion I don't think the todo is needed:
```suggestion
```

Reply:
Done